Add README for internal webhooks (#11)
sourcegraph/bitbucket-server-plugin: src/main/java/com/sourcegraph/webhook/README.md

# Webhooks

> The Sourcegraph Bitbucket plugin provides an internal implementation for webhooks that supports listening to events at a global or project scope.

Reviewer: nit:
```suggestion
The Sourcegraph Bitbucket Server plugin provides an internal implementation for webhooks that supports listening to events at a global or project scope.
```

## Contents
- [Payload](#payload)
- [REST API](#rest-api)
    * [List](#list)
    * [Create](#create)
    * [Delete](#delete)

## Payload

> The event payload request is sent when an event is fired. There will be five attempts before the request fails. This will be recorded in the logs.

Reviewer: >This will be recorded in the logs.

Does "This" refers to errors/failures only or every request?

Author: Successful requests are outputted as debug messages while failures will show up as an error. I added some clarification to the README.

> The request has these headers:

Reviewer: ```suggestion
Each request has the following headers:
```

> - `X-Event-Key` - contains the specific event key

Reviewer: ```suggestion
- `X-Event-Key` - The name of the event
```

> - `X-Hook-ID` - webhook id

Reviewer: ```suggestion
- `X-Hook-ID` - The ID that represents the webhook
```

> - `X-Hook-Name` - webhook name

Reviewer: ```suggestion
- `X-Hook-Name` - The name of the webhook that sends the request
```

- `X-Hub-Signature` - HMAC SHA256 hash of the payload (request body)

> The payload is stored in the request body as a JSON object. They follow the schema specified in [Event Payload](https://confluence.atlassian.com/bitbucketserver0516/event-payload-966061436.html?utm_campaign=in-app-help&utm_medium=in-app-help&utm_source=stash#Eventpayload-repositoryevents) and

Reviewer: ```suggestion
The payload is stored in the request body as a JSON object. Each payload follows the schema specified in [Event Payload](https://confluence.atlassian.com/bitbucketserver0516/event-payload-966061436.html?utm_campaign=in-app-help&utm_medium=in-app-help&utm_source=stash#Eventpayload-repositoryevents) and
```
nit: why is there a line break?

[Bitbucket Entities](https://docs.atlassian.com/bitbucket-server/docs/5.16.0/reference/javascript/JSON.html) for each event.

> Example (`pr:merged`):

Reviewer: ```suggestion
Here is an example payload for the event `pr:merged`:
```

> <img width="1266" alt="Screen Shot 2019-11-13 at 6 21 57 AM" src="https://user-images.githubusercontent.com/3507526/68772049-fc7a3480-05dd-11ea-9676-707b40fd3daf.png">

Reviewer: What do you think about pasting the actual JSON here? Image is very hard to copy.

## REST API
Interacting with the webhook REST endpoints requires [authentication](https://developer.atlassian.com/server/bitbucket/how-tos/example-basic-authentication/) from a system admin account.

#### List
```
$ curl -X GET 'https://${BITBUCKET_SERVER_URL}/rest/sourcegraph-admin/1.0/webhook'
```
There will be a 200 response code if successful.

Output: JSON serialized `[]Webhook`  

> Ex: `[{"id":1,"name":"test","scope":"global","identifier":"","events":["pr"],"endpoint":"endpoint","secret":"secret"}]`

Reviewer: **Suggestion**:

Output:
```json
[
   {
      "id":1,
      "name":"test",
      "scope":"global",
      "identifier":"",
      "events":[
         "pr"
      ],
      "endpoint":"endpoint",
      "secret":"secret"
   }
]
```

#### Create
```
curl -X POST 'https://${BITBUCKET_SERVER_URL}/rest/sourcegraph-admin/1.0/webhook' \
    -H 'Content-Type: application/json' \

>     -d '{"name":"", "scope":"", "identifier":"", "events":["", "", ...], "endpoint":"", "secret":""}'

Reviewer: I think it would be much easier for reader if `...` are real/usable events:
```suggestion
    -d '{"name":"", "scope":"", "identifier":"", "events":["pr", "pr:opened", "pr:modified"], "endpoint":"", "secret":""}'
```

# Webhooks

> The Sourcegraph Bitbucket Server plugin provides an internal implementation for webhooks that supports listening to events at a global or project scope.

Reviewer 2: ```suggestion
The Sourcegraph Bitbucket Server plugin provides an implementation for webhooks that supports listening to events at a global or project scope.
```

## Contents
- [Payload](#payload)
- [REST API](#rest-api)
    * [List](#list)
    * [Create](#create)
    * [Delete](#delete)

## Payload
The event payload HTTP POST request is sent when an event is fired. In the case that the request fails, there will be four further attempts with an interval of 10 seconds before it stops trying. The outcomes of these requests are recorded in the logs.

Each request has the following headers:
- `X-Event-Key` - The name of the event
- `X-Hook-ID` - The ID that represents the webhook
- `X-Hook-Name` - The name of the webhook that sends the request
- `X-Hub-Signature` - HMAC SHA256 hash of the payload (request body)

> The payload is stored in the request body as a JSON object. Each payload follows the schema specified in [Event Payload](https://confluence.atlassian.com/bitbucketserver0516/event-payload-966061436.html?utm_campaign=in-app-help&utm_medium=in-app-help&utm_source=stash#Eventpayload-repositoryevents) and [Bitbucket Entities](https://docs.atlassian.com/bitbucket-server/docs/5.16.0/reference/javascript/JSON.html).

Reviewer 2: ```suggestion
The payload is sent in the request body as a JSON object. Each payload follows the schema specified in [Event Payload](https://confluence.atlassian.com/bitbucketserver0516/event-payload-966061436.html?utm_campaign=in-app-help&utm_medium=in-app-help&utm_source=stash#Eventpayload-repositoryevents) and [Bitbucket Entities](https://docs.atlassian.com/bitbucket-server/docs/5.16.0/reference/javascript/JSON.html).
```

Here is an example payload for the event `pr:merged`:
<img width="1266" alt="Screen Shot 2019-11-13 at 6 21 57 AM" src="https://user-images.githubusercontent.com/3507526/68772049-fc7a3480-05dd-11ea-9676-707b40fd3daf.png">

## REST API
Interacting with the webhook REST endpoints requires [authentication](https://developer.atlassian.com/server/bitbucket/how-tos/example-basic-authentication/) from a system admin account.

#### List
```
$ curl -X GET 'https://${BITBUCKET_SERVER_URL}/rest/sourcegraph-admin/1.0/webhook'
```
There will be a 200 response code if successful.

Output: JSON serialized `[]Webhook`  
```
[{
    "id": "1",
    "name": "webhook",
    "scope": "global",
    "identifier": "",
    "events": ["pr"],

>     "endpoint": "endpoint",

Reviewer 2: ```suggestion
    "endpoint": "https://${SOURCEGRAPH_URL}/.api/bitbucketserver-webhooks",
```

    "secret": "secret"
}]
```

#### Create
```
curl -X POST 'https://${BITBUCKET_SERVER_URL}/rest/sourcegraph-admin/1.0/webhook' \
    -H 'Content-Type: application/json' \

>     -d '{"name":"", "scope":"", "identifier":"", "events":["pr:opened", "pr:merged", "pr:comment:added"], "endpoint":"", "secret":""}'

Reviewer 2: What sort of validation does this endpoint do? What responses can we expect when validation fails?

Author: At the moment, there is minimal validation.
```
401 - Unauthorized requests
404 - No such project or repository
422 - Invalid selector
```

Reviewer 2: Please document these, and what the response payload is.

```
There will be a 204 response code if successful.  

**JSON Fields:**
* name - string id for webhook
* scope - can be either `global`, `project`, or `repository`

> * identifier - empty (`""`) if scope is global, otherwise, name of `project` or `repository`

Reviewer 2: I didn't catch this in the other PR. Seems weird to have to specify these two fields (i.e. `scope` AND `identifier`). Could we somehow merge `identifier` into `scope`?

Author: Yep, that makes more sense!

Author: I opened PR #12 for this change. I'll update these docs.
